Suppress diff on the skip_confirmation attribute of gitlab_user

PatrickRice-KSC opened the pull request:
## Description

Fix an issue with gitlab_user where the `skip_confirmation` attribute results in a non-empty plan on every run.

Closes #1101

### PR Checklist Acknowledgement

<!-- For a smooth review process, please run through this checklist before submitting a PR, and check the box when done. -->

- [x] I acknowledge that all of the following items are true, where applicable:
  - Resource attributes match 1:1 the names and structure of the API resource in [the GitLab API documentation](https://docs.gitlab.com/ee/api/).
  - [Examples](https://github.com/gitlabhq/terraform-provider-gitlab/tree/main/examples) are updated with:
    - A \*.tf file for the resource/s with at least one usage example
    - A \*.sh file for the resource/s with an import example (if applicable)
  - New resources have at minimum a basic test with three steps:
    - Create the resource
    - Update the attributes
    - Import the resource
  - No new `//lintignore` comments were copied from existing code. (Linter rules are meant to be enforced on new code.)

PatrickRice-KSC commented:
@timofurrer  - I'm not really proficient in ruby, so I can't speak to how possible it is to expose the API via gitlab 🙂 . 

This is a problem with more than just import though, it'll happen any time a value is set into that field in a config, because we never set the field into state anywhere (likely because it's not returned from any of the APIs). You can see that code here: https://github.com/gitlabhq/terraform-provider-gitlab/blob/main/internal/provider/resource_gitlab_user.go#L123

We could potentially set it into state in create/update and assert that the value got set appropriately during insert/update, but that seems a bit of an anti-pattern since we've seen instances before where the GitLab API returned 200 and didn't actually set the value, so we wouldn't want to give the impression that an operation succeeded when it didn't.  I debated seeing if we could use `getChange`, but that compares state to value, which means since the state is always nil it still comes back with a change. I also debated whether there was something more elegant than just `return true` for the skip func, but it's all comparing to existing state which would be nil  :-/

timofurrer commented:
> This is a problem with more than just import though, it'll happen any time a value is set into that field in a config, because we never set the field into state anywhere (likely because it's not returned from any of the APIs).

> We could potentially set it into state in create/update and assert that the value got set appropriately during insert/update,

Yes, I guess we meant the same here (I was just assume that and wasn't specific enough, sorry) ... basically: 

It would require that we set the `skip_confirmation` attribute from the config to state during create / update - similar to what we do with a `token`. In the `update` we would just set the value from the config every time (because we don't know if there was a drift).

Then, there is still the import ... for which we may just use what's in the config ...

I'll dig into GitLab once I have some time.

olhado commented:
Just poking this so it stays alive. The lifecycle `ignore_changes` workaround is not great.

PatrickRice-KSC commented:
@timofurrer  - Should we just update this to set the state on create/update, and skip import verification for now? It would require a separate attribute to change in order to trigger an "update" operation because we'd still need the skip func, but at least it's slightly more graceful.

timofurrer commented:
@PatrickRice-KSC I still didn't have any time to dig into the GitLab side of things ... 

> Should we just update this to set the state on create/update, and skip import verification for now?

Yes, I'd skip import for now ..

PatrickRice-KSC commented:
@timofurrer  - I updated the way this is handled to be set directly from the config on create/update, removed the suppress function and set the attribute as computed. 

I also updated the test to check the value directly since the attribute doesn't exist on the *gitlab.User object (since it can't be retrieved from any API).

Shocktrooper commented:
@PatrickRice-KSC Is this good for a review?

timofurrer commented on internal/provider/resource_gitlab_user.go:
```diff
@@ -74,11 +74,15 @@ var _ = registerResource("gitlab_user", func() *schema.Resource {
-				Default:     true,
-				ForceNew:    true,
+				Description: `Boolean, defaults to true. Whether to skip confirmation.
+				
```
The rendering of this is not really nice:

![image](https://user-images.githubusercontent.com/1008252/190596873-1c0f57cb-3a99-4bdf-9284-14a52a4fe630.png)

Do you mind removing the blank line. It would change it to:

![image](https://user-images.githubusercontent.com/1008252/190596994-84111968-a4f7-422a-9da9-8d41ce183c82.png)

```suggestion
```

Shocktrooper replied:
@PatrickRice-KSC

PatrickRice-KSC replied:
Wow, I'm not sure how I missed this comment for over a month. Jeez, sorry guys (and thanks for the reminder @Shocktrooper !)

I'll get those changes made so we can get this merged 😄

Shocktrooper replied:
No worries 😄 I was just like huh I don't think there's anything else to do besides this review comment

timofurrer reviewed (approved):
@PatrickRice-KSC 

IHMO, we could actually merge it given it provides some value for some people already 😊 

However, what would you think about just suppressing any change on this field?

Like we do here with the `skip_wait_for_default_branch_protection` attribute:

https://github.com/gitlabhq/terraform-provider-gitlab/blob/2aa1c94e80d23d2e668330ea7b0e9402cc2edc09/internal/provider/resource_gitlab_project.go#L683

PatrickRice-KSC commented:
> IHMO, we could actually merge it given it provides some value for some people already 😊
> 
> However, what would you think about just suppressing any change on this field?
> 
> Like we do here with the `skip_wait_for_default_branch_protection` attribute:

Hey @timofurrer  - I'm working through fixing the test failures, and when we set the DiffSuppressFunc to just return `true`, the change no longer works how we want it to, because it suppresses the initial _creation_ diff as well, and the value _always returns false_. So I think we need to leave the DiffSuppressFunc off, and just let it set the value into state on the first apply. Thoughts?
